### PR TITLE
votca 2022.1 (new formula)

### PR DESCRIPTION
**Pull request by junghans**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by junghans** [comment]
@mabraham @eirrgang @pszi1ard do you understand this linking error below? (On Linux I cannot reproduce this)
```
Undefined symbols for architecture arm64:
  "read_first_frame(gmx_output_env_t const*, t_trxstatus**, std::__1::__fs::filesystem::path const&, t_trxframe*, int)", referenced from:
      votca::csg::GMXTrajectoryReader::FirstFrame(votca::csg::Topology&) in gmxtrajectoryreader.cc.o
  "open_trx(std::__1::__fs::filesystem::path const&, char const*)", referenced from:
      votca::csg::GMXTrajectoryWriter::Open(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool) in gmxtrajectorywriter.cc.o
  "read_tpx(std::__1::__fs::filesystem::path const&, t_inputrec*, float (*) [3], int*, float (*) [3], float (*) [3], gmx_mtop_t*)", referenced from:
      votca::csg::GMXTopologyReader::ReadTopology(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, votca::csg::Topology&) in gmxtopologyreader.cc.o
ld: symbol(s) not found for architecture arm64
```

**Comment by eirrgang** [comment]
> @mabraham @eirrgang @pszi1ard do you understand this linking error below? (On Linux I cannot reproduce this)
> 
> ```
> Undefined symbols for architecture arm64:
>   "read_first_frame(gmx_output_env_t const*, t_trxstatus**, std::__1::__fs::filesystem::path const&, t_trxframe*, int)", referenced from:
>       votca::csg::GMXTrajectoryReader::FirstFrame(votca::csg::Topology&) in gmxtrajectoryreader.cc.o
>   "open_trx(std::__1::__fs::filesystem::path const&, char const*)", referenced from:
>       votca::csg::GMXTrajectoryWriter::Open(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool) in gmxtrajectorywriter.cc.o
>   "read_tpx(std::__1::__fs::filesystem::path const&, t_inputrec*, float (*) [3], int*, float (*) [3], float (*) [3], gmx_mtop_t*)", referenced from:
>       votca::csg::GMXTopologyReader::ReadTopology(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, votca::csg::Topology&) in gmxtopologyreader.cc.o
> ld: symbol(s) not found for architecture arm64
> ```

Unintentional cross-compiling? It looks like votca is compiling for ARM, but the libgromacs was for x86_64 (or whatever)?

**Comment by junghans** [comment]
@eirrgang I think I got it, in https://github.com/gromacs/gromacs/commit/97fe261c29c6c6f9dbe912bf30f267b5ca20d6f7#diff-bb6d875dcd708aadb3095833f8648a0e521eee03f9507ebe01136863707367eeR233 the signature of `read_first_frame()` was changed. I assume on OSX/ARM64 `std::filesystem::path` is not the same as `char *`, will try to fix this in https://github.com/votca/votca/pull/1059.

**Comment by eirrgang** [comment]
If https://cmake.org/cmake/help/latest/variable/CMAKE_OSX_ARCHITECTURES.html is set when configuring gromacs, it should get set in `share/cmake/gromacs*/gromacs*-hints.cmake` to the same value. But I'm not sure if there is a reasonable way to derive an appropriate value. We should probably avoid writing that line if it is an empty string, though.

Really, it seems like if it is truly an architecture difference, CMake should be propagating some sort of relevant transitive usage requirement with the `target_link_libraries(Gromacs::libgromacs)`. Maybe our export script needs an update.

If this turns out to be an architecture mismatch rather than a loader path toolchain mismatch, consider opening a gromacs issue suggesting stronger hints through the Gromacs::libgromacs IMPORTED_TARGET or the hints file, or possibly a warning in the CMake config module.

**Comment by eirrgang** [comment]
> @eirrgang I think I got it, in [gromacs/gromacs@97fe261#diff-bb6d875dcd708aadb3095833f8648a0e521eee03f9507ebe01136863707367eeR233](https://github.com/gromacs/gromacs/commit/97fe261c29c6c6f9dbe912bf30f267b5ca20d6f7#diff-bb6d875dcd708aadb3095833f8648a0e521eee03f9507ebe01136863707367eeR233) the signature of `read_first_frame()` was changed. I assume on OSX/ARM64 `std::filesystem::path` is not the same as `char *`, will try to fix this in [votca/votca#1059](https://github.com/votca/votca/pull/1059).

Oh, you borrowed the declaration from an old header and just need to update it?

**Comment by junghans** [comment]
> > @eirrgang I think I got it, in [gromacs/gromacs@97fe261#diff-bb6d875dcd708aadb3095833f8648a0e521eee03f9507ebe01136863707367eeR233](https://github.com/gromacs/gromacs/commit/97fe261c29c6c6f9dbe912bf30f267b5ca20d6f7#diff-bb6d875dcd708aadb3095833f8648a0e521eee03f9507ebe01136863707367eeR233) the signature of `read_first_frame()` was changed. I assume on OSX/ARM64 `std::filesystem::path` is not the same as `char *`, will try to fix this in [votca/votca#1059](https://github.com/votca/votca/pull/1059).
> 
> Oh, you borrowed the declaration from an old header and just need to update it?

This only happened in the 2023 release.

**Comment by eirrgang** [comment]
> This only happened in the 2023 release.

I'm just struggling to understand why the signature mismatch wasn't caught until the linking phase. Why didn't the compiler catch it?

**Comment by junghans** [comment]
> > This only happened in the 2023 release.
> 
> I'm just struggling to understand why the signature mismatch wasn't caught until the linking phase. Why didn't the compiler catch it?

Yes, I don't get that either.

Now, with my patch, I get a slightly different error:
```
Undefined symbols for architecture arm64:
  "read_first_frame(gmx_output_env_t const*, t_trxstatus**, std::__1::__fs::filesystem::path const&, t_trxframe*, int)", referenced from:
      votca::csg::GMXTrajectoryReader::FirstFrame(votca::csg::Topology&) in gmxtrajectoryreader.cc.o
  "open_trx(std::__1::__fs::filesystem::path const&, char const*)", referenced from:
      votca::csg::GMXTrajectoryWriter::Open(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool) in gmxtrajectorywriter.cc.o
  "read_tpx(std::__1::__fs::filesystem::path const&, t_inputrec*, float (*) [3], int*, float (*) [3], float (*) [3], gmx_mtop_t*)", referenced from:
      votca::csg::GMXTopologyReader::ReadTopology(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, votca::csg::Topology&) in gmxtopologyreader.cc.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

The thing that is stange is the signature, why is it:
```
read_first_frame(gmx_output_env_t const*, t_trxstatus**, std::__1::__fs::filesystem::path const&, t_trxframe*, int)
```
and not
```
read_first_frame(gmx_output_env_t const*, t_trxstatus**,std::filesystem::path const&,t_trxframe*, int)
```

Stack overlow suggest this is an issue with mixing `-stdlib=libc++` (LLVM C++ runtime) and `-stdlib=libstdc++` (GNU C++ runtime):
https://stackoverflow.com/questions/29293394/where-does-the-1-symbol-come-from-when-using-llvms-libc

**Comment by eirrgang** [comment]
> I'm just struggling to understand why the signature mismatch wasn't caught until the linking phase. Why didn't the compiler catch it?

Ah. I guess it is because the STL `path` has layers of templated creation functions and the compiler doesn't need a definition for the constructor implied by the function argument that needs to be converted.

**Comment by eirrgang** [comment]
> Stack overlow suggest this is an issue with mixing -stdlib=libc++ (LLVM C++ runtime) and -stdlib=libstdc++ (GNU C++ runtime)

Exactly.

The different standard library implementations use different ABI-specific / implementation-specific namespaces for the underlying details. It's why we should avoid standard library types in our API to improve ABI compatibility.

**Comment by junghans** [comment]
Hmm I think gromacs does something non-standard in `cmake/FindLibStdCpp.cmake`, which makes the gromacs' bottle being built with a different `stdlib` than brew's default.

@p-linnane is there a way to look up the logs of the latest gromacs bottle build?

**Comment by eirrgang** [comment]
If you use the CMAKE_OSX_DEPLOYMENT_TARGET from the gromacs-hints.cmake file, it should imply a matching standard library implementation.

Unfortunately, the CMake tooling for this isn't totally unified for OSX vs. non-OSX (and [progress is slow](https://gitlab.kitware.com/cmake/cmake/-/issues/20851)).

**Comment by p-linnane** [comment]
@junghans https://github.com/Homebrew/homebrew-core/actions/runs/4771067654

**Comment by eirrgang** [comment]
See also https://gitlab.com/gromacs/gromacs/-/issues/4498

**Comment by junghans** [comment]
> @junghans https://github.com/Homebrew/homebrew-core/actions/runs/4771067654

@p-linnane I see where it spends 25min to build gromacs, but where is the output of the build process?

**Comment by junghans** [comment]
> If the gromacs homebrew recipe is not explicitly setting CMAKE_OSX_DEPLOYMENT_TARGET, I recommend doing so, and using the same value for dependent packages (such as by using the hints file or through some other mechanism, if brew has a system for managing such things).

It does not, can you make a PR?

**Comment by p-linnane** [comment]
The logs are not more detailed than that since the build was successful. If you'd like to generate full build logs locally run `brew install --verbose --build-from-source gromacs` and that should get you what you need.

**Comment by eirrgang** [comment]
> > If the gromacs homebrew recipe is not explicitly setting CMAKE_OSX_DEPLOYMENT_TARGET, I recommend doing so, and using the same value for dependent packages (such as by using the hints file or through some other mechanism, if brew has a system for managing such things).
> 
> It does not, can you make a PR?

I wouldn't know where to look. I've noticed that a lot of the gromacs dependencies set a non-default platform target that tends to track the host environment, so homebrew must have some generic mechanism for this and I don't know what it is. I also don't know whether it is appropriate or valid for one homebrew recipe to depend on an installation artifact from another build recipe for configuring its build system (i.e. using the installed hints file) or whether there is a better way of establishing compatible toolchains.

If it has not been resolved in a few weeks, I may be able to investigate further and try to contribute.

**Comment by carlocab** [comment]
> > @junghans https://github.com/Homebrew/homebrew-core/actions/runs/4771067654
> 
> @p-linnane I see where it spends 25min to build gromacs, but where is the output of the build process?

Logs are available as artifacts. E.g.:
```sh
gh run download 4771067654 --pattern 'logs*' -R Homebrew/homebrew-core
```

**Comment by carlocab** [comment]
> Or mismatched toolchains that munge symbols differently / use different standard library implementations?

Yea, this is your issue. `gromacs` uses GNU `g++-*` to compile (not sure why), so will link  with `libstdc++`.

**Comment by eirrgang** [comment]
> Yea, this is your issue. `gromacs` uses GNU `g++-*` to compile (not sure why), so will link with `libstdc++`.

Historically, homebrew gcc was the only decent way to build with OpenMP support on OS X.

> ```
>  gcc = Formula["gcc"] 
>  cc = gcc.opt_bin/"gcc-#{gcc.any_installed_version.major}" 
> ```

Is this a pattern that votca could use to get its tool chain? Something like
```
gromacs = Formula["gromacs"]
hints = gromacs.opt_share/"cmake/gromacs/gromacs-hints.cmake"
```
and add `-C{hints}` to the cmake args?

**Comment by carlocab** [comment]
> Is this a pattern that votca could use to get its tool chain? Something like
> 
> ```
> 
> gromacs = Formula["gromacs"]
> 
> hints = gromacs.opt_share/"cmake/gromacs/gromacs-hints.cmake"
> 
> ```
> 
> and add `-C{hints}` to the cmake args?

Seems reasonable. This should do
```ruby
fails_with :clang
```
to indicate it intends to use GCC, though.

**Comment by eirrgang** [comment]
> > Is this a pattern that votca could use to get its tool chain? Something like
> > ```
> > 
> > gromacs = Formula["gromacs"]
> > 
> > hints = gromacs.opt_share/"cmake/gromacs/gromacs-hints.cmake"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and add `-C{hints}` to the cmake args?
> 
> Seems reasonable. This should do
> 
> ```ruby
> fails_with :clang
> ```
> 
> to indicate it intends to use GCC, though.

Would there be a more idiomatic/unified way to express the toolchain of one package in terms of another's? It seems like `votca` shouldn't have to say `fails_with: clang` based on the author's knowledge that the gromacs recipe is assumed to use gcc at the time of writing.

**Comment by carlocab** [comment]
> Would there be a more idiomatic/unified way to express the toolchain of one package in terms of another's?

No, unfortunately not.

**Comment by junghans** [comment]
Maybe we should try to build gromacs with clang as it has openmp support these days....

**Comment by junghans** [comment]
Building votca with gcc leads to some missing symbols now in boost:
```
Undefined symbols for architecture arm64:
  "__ZN5boost15program_options11to_internalERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE", referenced from:
      __ZN5boost15program_options25basic_command_line_parserIcEC4EiPKPKc in application.cc.o
  "__ZN5boost15program_options19options_descriptionC1ERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEjj", referenced from:
      __ZN5votca5tools11Application17AddProgramOptionsERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE in application.cc.o
      __ZN5votca5tools11ApplicationC4Ev in application.cc.o
  "__ZN5boost15program_options6detail7cmdline21set_additional_parserENS_9function1ISt4pairINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESA_ERKSA_EE", referenced from:
      __ZN5votca5tools11Application16ParseCommandLineEiPPc in application.cc.o
  "__ZN5boost15program_options6detail7cmdlineC2ERKSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS9_EE", referenced from:
      __ZN5boost15program_options25basic_command_line_parserIcEC4EiPKPKc in application.cc.o
  "__ZN5boost15program_optionslsERSoRKNS0_19options_descriptionE", referenced from:
      __ZN5votca5tools11Application12ShowHelpTextERSo in application.cc.o
ld: symbol(s) not found for architecture arm64
collect2: error: ld returned 1 exit status
make[2]: *** [tools/src/libtools/libvotca_to
```

**Comment by fxcoudert** [comment]
In which boost libraries are those defined?

File: Formula/votca.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  depends_on "eigen"
+  depends_on "expat"
+  depends_on "fftw"
+  depends_on "gcc" # for OpenMP
+  depends_on "gromacs"
+  depends_on "hdf5"
+
+  on_macos do
+    depends_on "libomp"
+  end
```
Comment (eirrgang) [inline Formula/votca.rb:30]: This should be removed, right? Per https://github.com/Homebrew/homebrew-core/pull/129648#issuecomment-1553839689, `libgomp` will be provided through `gcc`?

```suggestion
```

**Comment by junghans** [comment]
> Undefined symbols: In which boost libraries are those defined?

I piped the error message through `c++flit`:
```
Undefined symbols for architecture arm64:
  "boost::program_options::to_internal(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)", referenced from:
      boost::program_options::basic_command_line_parser<char>::basic_command_line_parser(int, char const* const*) in application.cc.o
  "boost::program_options::options_description::options_description(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int, unsigned int)", referenced from:
      votca::tools::Application::AddProgramOptions(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in application.cc.o
      votca::tools::Application::Application() in application.cc.o
  "boost::program_options::detail::cmdline::set_additional_parser(boost::function1<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>)", referenced from:
      votca::tools::Application::ParseCommandLine(int, char**) in application.cc.o
  "boost::program_options::detail::cmdline::cmdline(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&)", referenced from:
      boost::program_options::basic_command_line_parser<char>::basic_command_line_parser(int, char const* const*) in application.cc.o
  "boost::program_options::operator<<(std::basic_ostream<char, std::char_traits<char> >&, boost::program_options::options_description const&)", referenced from:
      votca::tools::Application::ShowHelpText(std::basic_ostream<char, std::char_traits<char> >&) in application.cc.o
ld: symbol(s) not found for architecture arm64
collect2: error: ld returned 1 exit status
make[2]: *** [tools/src/libtools/libvotca_to
```

Only linux these symbols are part of `libboost_program_options.so`.

File: Formula/votca.rb (hunk 1 above)
Comment (junghans) [reply Formula/votca.rb]: Well, if you want to build with clang, you will still need this block.

**Comment by cho-m** [comment]
Taking first symbol as example, I am guessing missing
```
__ZN5boost15program_options11to_internalERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
```
should be similar to `libboost_program_options.dylib` symbol:
```
__ZN5boost15program_options11to_internalERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE
```
given that it demangles to:
```
boost::program_options::to_internal(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
```

---

I think the issue would be mixing C++ implementations and hitting ABI incompatibility.

**Comment by junghans** [comment]
Yeah I am thinking the same!

The Gromacs dependency forces us to use gcc, but the boost dependency was built with clang.

**Comment by junghans** [comment]
> On a second thought maybe the best right now is tobuild VOTCA without gromacs support.

That worked, I will add gromacs back in the mix later!

**Comment by junghans** [comment]
I am really not sure how to fix this:
```
Full audit votca --online --new-formula output
  votca
    * GitHub repository not notable enough (<30 forks, <30 watchers and <75 stars)
```

Most of our users aren't much on GitHub, as we provide rpm, deb, spack, Gentoo, conda and docker packages for many years, so stars on GitHub is usually not our metric.

**Comment by junghans** [comment]
> I am really not sure how to fix this:
> 
> ```
> Full audit votca --online --new-formula output
>   votca
>     * GitHub repository not notable enough (<30 forks, <30 watchers and <75 stars)
> ```
> 
> Most of our users aren't much on GitHub, as we provide rpm, deb, spack, Gentoo, conda and docker packages for many years, so stars on GitHub is usually not our metric.

Thanks to our users we passed the threshold!

**Comment by junghans** [comment]
@chenrui333 could you please review this PR as well?

File: Formula/votca.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      "-DCMAKE_DISABLE_FIND_PACKAGE_GROMACS=ON",
+      "-DENABLE_RPATH_INJECT=ON",
+    ]
+    system "cmake", "-S", ".", "-B", "build", *std_cmake_args, *args
+    system "cmake", "--build", "build"
+    system "cmake", "--install", "build"
+  end
+
+  test do
+    system "#{bin}/csg_property", "--help"
```
Comment (fxcoudert) [inline Formula/votca.rb:38]: Can we do something more clever here? Run a small analysis (maybe a trivial xyz file?)

File: Formula/votca.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
+class Votca < Formula
+  desc "Versatile Object-oriented Toolkit for Coarse-graining Applications"
+  homepage "https://www.votca.org/"
+  url "https://github.com/votca/votca/archive/refs/tags/v2023-rc.1.tar.gz"
```
Comment (fxcoudert) [inline Formula/votca.rb:4]: We do not accept pre-release versions.

**Review by fxcoudert — CHANGES_REQUESTED** [review]

File: Formula/votca.rb (hunk 2 above)
Comment (junghans) [reply Formula/votca.rb:39]: I added running `ctest` above that will run the whole test suite.

File: Formula/votca.rb (hunk 3 above)
Comment (junghans) [reply Formula/votca.rb:4]: Ok, I will push out the real release next week.

File: Formula/votca.rb (hunk 2 above)
Comment (fxcoudert) [reply Formula/votca.rb:39]: No, please remove `ctest` and add a small functional test in the `test do` block. See for example `lammps`

Comment (junghans) [reply Formula/votca.rb:39]: Ok, I added another test here.

**Comment by fxcoudert** [comment]
OK, the new test is passing, and the PR is overall fine. Before it can be accepted:
- migrate it to the latest stable version
- please squash all commits into a single commit, titled `votca 2023 (new formula)` (using the latest version)

**Comment by junghans** [comment]
@fxcoudert all done!

**Review by fxcoudert — APPROVED** [review]

**Pull request merged**